src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11800,6 +11869,77 @@ private ExpressionSyntax ParseNewExpression()
+                static @this => @this.ParseCollectionElement(),
+                skipBadCollectionElementTokens,
+                allowTrailingSeparator: true,
+                requireOneElement: false,
+                allowSemicolonAsSeparator: false);
+
+            return _syntaxFactory.CollectionCreationExpression(
+                openBracket,
+                list,
+                this.EatToken(SyntaxKind.CloseBracketToken));
+
+            static PostSkipAction skipBadCollectionElementTokens(
+                LanguageParser @this, ref SyntaxToken openBracket, SeparatedSyntaxListBuilder<CollectionElementSyntax> list, SyntaxKind expectedKind, SyntaxKind closeKind)
+            {
+                return @this.SkipBadSeparatedListTokensWithExpectedKind(ref openBracket, list,
+                    static p => p.CurrentToken.Kind != SyntaxKind.CommaToken && !p.IsPossibleCollectionElement(),
+                    static (p, closeKind) => p.CurrentToken.Kind == closeKind,
+                    expectedKind, closeKind);
+            }
+        }
+
+        private bool IsPossibleCollectionElement()
+        {
+            if (this.IsPossibleExpression())
+                return true;
+
+            // Checking for ':' is for error recovery when someone has a dictionary element and is missing the key part.
+            if (this.CurrentToken.Kind == SyntaxKind.ColonToken)
+                return true;
+
+            // Checking for `keyword:` is for error recovery when typing a dictionary element, but the partial
+            // identifier happens to match a keyword.
+            if (SyntaxFacts.IsReservedKeyword(this.CurrentToken.Kind) && this.PeekToken(1).Kind == SyntaxKind.ColonToken)
+                return true;
+
+            return false;
+        }
+
+        private CollectionElementSyntax ParseCollectionElement()
+        {
+            var dotDotToken = this.TryEatToken(SyntaxKind.DotDotToken);
+            if (dotDotToken != null)
+                return _syntaxFactory.SpreadElement(dotDotToken, this.ParseExpressionCore());
+
+            // Be resilient to `keyword:val` if the user hits that while typing out a full identifier.
+            ExpressionSyntax expression;
+            if (SyntaxFacts.IsReservedKeyword(this.CurrentToken.Kind) && this.PeekToken(1).Kind == SyntaxKind.ColonToken)
+            {
+                var keyword = this.EatTokenWithPrejudice(SyntaxKind.IdentifierToken);
+                var identifier = ConvertToIdentifier(keyword).WithDiagnosticsGreen(keyword.GetDiagnostics());
+                expression = _syntaxFactory.IdentifierName(identifier);
+            }
+            else
+            {
+                expression = this.ParseExpressionCore();
+            }
+
+            var colonToken = this.TryEatToken(SyntaxKind.ColonToken);
+            return colonToken != null
+                ? _syntaxFactory.DictionaryElement(expression, colonToken, this.ParseExpressionCore())
```

Comment:
Spec question: I expected a dictionary element (and thus syntax model) to be `name ':' expression` but it appears to be `expression ':' expression`. What is the rationale?

Reply:
because this is for dictionaries, and keys in dictionaries can be arbitrary expressions :)

Reply:
Added a bullet to test plan. The colon is normally used with a name. A syntax like `[key] = value` would be heavier, but would fit better with precedent.

Reply:
Note: the working group considered this and explicitly rejected it.  Not following precedent is a core design pivot here as the precedent we have in so many places is far too heavyweight.  We can certainly let the LDM weigh in, but this was an explicit syntactic choice considered (exactly against existing forms like the one suggested).